spec/API_specification/constants.md

```diff
@@ -2,7 +2,9 @@
 
 > Array API specification for constants.
 
-A conforming implementation of the array API standard must provide and support the following constants.
+A conforming implementation of the array API standard must provide and support the following constants adhering to the following conventions.
+
+-   Each constant must have a Python numeric data type (i.e., `int`, `float`, or `complex`) and be provided as a Python scalar value.
```

Comment:
Shouldn't they all specifically be `float`?

Reply:
Also, any reason to not just require exact values for `pi` and `e`, like

```py
pi = 3.141592653589793
e = 2.718281828459045
nan = float('nan')
inf = float('inf')
```

Reply:
The descriptions for each constant state the required Python numeric data type. In general, as a matter of convention, a constant must be a Python scalar. This section is intended to cover both current and potential use cases where constants may be either `float` or `int`.

Reply:
You mean if a library includes constants that are outside of the spec? Or do you mean for potential future versions of the spec? Including `int` and `complex` here just seems confusing to me since all four constants are floats.

Reply:
Future versions.

I am not sure the source of the confusion. The general principle states that a constant must have a numeric data type and then, for each constant, we specify the exact data type for that constant. The spec, as currently written, does not say that constants must be provided as an `int`, `float`, AND `complex`. Nor is the option available to choose which type is desired, given that the descriptions for each of the current constants state that they must be IEEE 754 floats.

Reply:
Regarding exact values for `pi` and `e`, to do so, we'd need to specify the exact bit sequence. From C11 6.4.4.2/3

> When `FLT_RADIX` is not a power of `2`, the result is either the nearest representable value, or the larger or smaller representable value immediately adjacent to the nearest representable value, chosen in an implementation-defined manner.

Reply:
How the constants are shown in the spec follows NumPy's [docs](https://numpy.org/doc/stable/reference/constants.html#numpy.e).

Reply:
I agree with @asmeurer here, if we have 4 constants that are all `float`, it makes little sense to include `int` and `complex`. It's also not future-proof; the only non-`float` constant in numpy is an alias for `None`, so not numerical.

Reply:
That's the C spec, but in Python floats exactly round-trip from their string representation (though I'm actually having a hard time finding documentation of this fact). However, we can also specify a hex representation using (see `float.hex` and `float.fromhex`).

Reply:
@asmeurer Round tripping from a string representation is not guaranteed due to floating-point...

```
In [1]: float("3.141592653589793")                                                                                                          
Out[1]: 3.141592653589793

In [2]: float("3.141592653589792")                                                                                                          
Out[2]: 3.141592653589792

In [3]: float("3.1415926535897925")                                                                                                         
Out[3]: 3.1415926535897927

In [4]: float("3.1415926535897932")                                                                                                         
Out[4]: 3.141592653589793
```

However, going the other way should round trip (i.e., map to the same bit sequence), but this is tricky when the value is not already rounded as shown below...

```
In [5]: str(float("3.1415926535897925"))                                                                                                    
Out[5]: '3.1415926535897927'

In [6]: float(str(3.1415926535897927))                                                                                                      
Out[6]: 3.1415926535897927

In [7]: str(3.1415926535897925)                                                                                                             
Out[7]: '3.1415926535897927'

In [8]: float(str(3.1415926535897925))                                                                                                      
Out[8]: 3.1415926535897927
```

The point regarding the C spec is that the last digit can vary depending on the platform (i.e., is implementation-defined). I don't think Python is exceptional here.

The reason for saying **an** IEEE 754 floating-point representation of `3.1415926535897932384626433...` is to acknowledge that the last bit may not be the same across all platforms.

Reply:
Updated language in latest commits.